File: file_bonanza-functions/read_parse_files.js
import * as fs from 'fs';
import * as xml2js from 'xml2js';
import * as csvParser from 'csv-parser';
import * as yaml from 'js-yaml';

function readTextFile(filename, callback) {
    fs.readFile(filename, 'utf8', (err, data) => {
        if (err) {
            callback(err, null);
            return;
        }
        callback(null, data);
        
    });
}

function readAndParseXML(filename, callback) {
    fs.readFile(filename, 'utf8', (err, data) => {
        if (err) {
            callback(err, null);
            return;
        }

        xml2js.parseString(data, (parseErr, result) => {
            if (parseErr) {
                callback(parseErr, null);
                return;
            }
            
            const skills = result.skills;
            const cantrips = skills.cantrips[0].cantrip;
            const spells = skills.spells[0].spell;
            const wildshape = {
                animals: skills.wildshape[0].animal,
                elementals: skills.wildshape[0].elemental
            };

            const parsedData = {
                cantrips: cantrips,
                spells: spells,
                wildshape: wildshape
            };

            callback(null, parsedData);
        });
    });
}

function readAndParseJSON(filename, callback) {
    fs.readFile(filename, 'utf8', (err, data) => {
        if (err) {
            callback(err, null);
            return;
        }

        try {
            const jsonData = JSON.parse(data);
            callback(null, jsonData);
        } catch (parseErr) {
            callback(parseErr, null);
        }
    });
}

function readAndParseCSV(filename, callback) {
    const results = [];
    fs.createReadStream(filename)
        .pipe(csvParser())
        .on('data', (data) => results.push(data))
        .on('end', () => {
            callback(null, results);
        });
}

function readAndParseYAML(filename, callback) {
    fs.readFile(filename, 'utf8', (err, data) => {
        if (err) {
            callback(err, null);
            return;
        }

        try {
            const yamlData = yaml.load(data);
            callback(null, yamlData);
        } catch (parseErr) {
            callback(parseErr, null);
        }
    });
}

export { readTextFile, readAndParseXML, readAndParseJSON, readAndParseCSV, readAndParseYAML };

File: file_bonanza-functions/read_parse_files.py
import json
import xml.etree.ElementTree as ET
import csv
import yaml

def read_text_file(filename):
    try:
        with open(filename, 'r') as file:
            content = file.read()
            return content
    except FileNotFoundError:
        raise FileNotFoundError(f"Error: {filename} not found")
    except Exception as e:
        raise Exception(f"An error occurred: {e}")

def read_and_parse_xml(filename):
    try:
        tree = ET.parse(filename)
        root = tree.getroot()

        skills = root.find('skills')
        cantrips = [cantrip.text for cantrip in skills.find('cantrips')]
        spells = [spell.text for spell in skills.find('spells')]
        wildshape = {
            'animals': [animal.text for animal in skills.find('wildshape').find('animal')],
            'elementals': [elemental.text for elemental in skills.find('wildshape').find('elemental')]
        }

        parsed_data = {
            'cantrips': cantrips,
            'spells': spells,
            'wildshape': wildshape
        }

        return parsed_data
    except FileNotFoundError:
        raise FileNotFoundError(f"Error: {filename} not found")
    except Exception as e:
        raise Exception(f"An error occurred: {e}")

def read_and_parse_json(filename):
    try:
        with open(filename, 'r') as file:
            json_data = json.load(file)
            return json_data
    except FileNotFoundError:
        raise FileNotFoundError(f"Error: {filename} not found")
    except Exception as e:
        raise Exception(f"An error occurred: {e}")

def read_and_parse_csv(filename):
    try:
        with open(filename, 'r') as file:
            csv_data = list(csv.DictReader(file))
            return csv_data
    except FileNotFoundError:
        raise FileNotFoundError(f"Error: {filename} not found")
    except Exception as e:
        raise Exception(f"An error occurred: {e}")

def read_and_parse_yaml(filename):
    try:
        with open(filename, 'r') as file:
            yaml_data = yaml.safe_load(file)
            return yaml_data
    except FileNotFoundError:
        raise FileNotFoundError(f"Error: {filename} not found")
    except Exception as e:
        raise Exception(f"An error occurred: {e}")